src/Tribe/Aggregator/Service.php

```diff
@@ -0,0 +1,255 @@
+	 */
+	protected $api_key;
+
+	/**
+	 * Static Singleton Factory Method
+	 *
+	 * @return Tribe__Events__Aggregator__Service
+	 */
+	public static function instance( Tribe__Events__Aggregator $aggregator ) {
+		if ( ! self::$instance ) {
+			self::$instance = new self( $aggregator );
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Constructor
+	 */
+	public function __construct( $aggregator ) {
+		$this->aggregator = $aggregator;
+		$this->api_key = $this->aggregator->get_license_key();
+
+		if ( defined( 'EVENT_AGGREGATOR_API_BASE_URL' ) ) {
+			$this->api_base_url = EVENT_AGGREGATOR_API_BASE_URL;
+		}
+
+		$this->get_origins();
+	}
+
+	/**
+	 * Builds an endpoint URL
+	 *
+	 * @param string $endpoint Endpoint for the Event Aggregator service
+	 *
+	 * @return string
+	 */
+	public function build_url( $endpoint ) {
+		$url = "{$this->api_base_url}{$this->api_root}{$this->api_version}/{$endpoint}";
+		$url = add_query_arg( 'key', $this->api_key, $url );
+
+		return $url;
+	}
+
+	/**
+	 * Performs a GET request against the Event Aggregator service
+	 *
+	 * @param string $url Endpoint URL
+	 * @param array $data Array of parameters to send to the endpoint
+	 *
+	 * @return stdClass
+	 */
+	public function get( $endpoint, $data = array() ) {
+		$url = $this->build_url( $endpoint );
+		$url = esc_url_raw( add_query_arg( $data, $url ) );
+
+		$response = wp_remote_get( $url );
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		// if the response is not an image, let's json decode the body
+		if ( ! preg_match( '/image/', $response['headers']['content-type'] ) ) {
+			$response = json_decode( wp_remote_retrieve_body( $response ) );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Performs a POST request against the Event Aggregator service
+	 *
+	 * @param string $url Endpoint URL
+	 * @param array $data Array of parameters to send to the endpoint
+	 *
+	 * @return stdClass
+	 */
+	public function post( $endpoint, $data = array() ) {
+		$url = $this->build_url( $endpoint );
+
+		if ( empty( $data['body'] ) ) {
+			$args = array( 'body' => $data );
+		} else {
+			$args = $data;
+		}
+
+		$response = wp_remote_post( $url, $args );
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response = json_decode( wp_remote_retrieve_body( $response ) );
+
+		return $response;
+	}
+
+	/**
+	 * Fetch origins from service
+	 *
+	 * @return array
+	 */
+	public function get_origins() {
+		$origins = array(
+			(object) array(
+				'id' => 'csv',
+				'name' => __( 'CSV File', 'the-events-calendar' ),
+			),
+		);
+
+		// if the user doesn't have a license key, don't bother hitting the service
+		if ( ! $this->api_key ) {
+			return $origins;
+		}
+
+		$response = $this->get( 'origin' );
+
+		if ( $response && 'success' === $response->status ) {
+			$origins = array_merge( $origins, $response->data->origin );
+		}
+
+		return $origins;
+	}
+
+	/**
+	 * Fetch import data from service
+	 *
+	 * @return array
```

Comment:
This method can return a `WP_Error` or the result of `$response` (which is returns a `stdClass` object or `WP_Error` only, unless I'm misreading).